core/shared/src/main/scala/zio/Schedule.scala

```diff
@@ -493,23 +516,48 @@ trait Schedule[-Env, -In, +Out] extends Serializable { self =>
    */
   def intersectWith[Env1 <: Env, In1 <: In, Out2](
     that: Schedule[Env1, In1, Out2]
-  )(
-    f: (Interval, Interval) => Interval
-  )(implicit zippable: Zippable[Out, Out2]): Schedule.WithState[(self.State, that.State), Env1, In1, zippable.Out] =
+  )(f: (Interval, Interval) => Interval)(implicit
+    zippable: Zippable[Out, Out2]
+  ): Schedule.WithState[(self.State, that.State), Env1, In1, zippable.Out] =
     new Schedule[Env1, In1, zippable.Out] {
       type State = (self.State, that.State)
-      val initial: State =
-        (self.initial, that.initial)
+      val initial = (self.initial, that.initial)
+
+      def loop(
```

Comment:
We recurse here until either schedule is done or we find a nonempty interval when both schedules are willing to recur. There is some question here of whether this should be the responsibility of the schedule or the driver but I think this makes the most sense because there are some ways of running the schedule other that the driver and otherwise we need to keep the last output in the state which will create variance issues.

Reply:
Interestingly there is a possibility of non-termination.